Imagine/Filter/PostProcessor/JpegOptimPostProcessor.php

```diff
@@ -46,129 +40,135 @@ class JpegOptimPostProcessor implements PostProcessorInterface
-            throw new \RuntimeException(sprintf('Temp file can not be created in "%s".', $tempDir));
-        }
+        $file = $this->writeTemporaryFile($binary, $options, 'imagine-post-processor-jpegoptim');
 
-        $processArguments = [$this->jpegoptimBin];
+        $arguments = $this->getProcessArguments($options);
+        $arguments[] = $file;
+        $process = $this->createProcess($arguments, $options);
+        $process->run();
 
-        $stripAll = \array_key_exists('strip_all', $options) ? $options['strip_all'] : $this->stripAll;
-        if ($stripAll) {
-            $processArguments[] = '--strip-all';
+        if (!$this->isSuccessfulProcess($process)) {
+            unlink($file);
+            throw new ProcessFailedException($process);
         }
 
-        $max = \array_key_exists('max', $options) ? $options['max'] : $this->max;
-        if ($max) {
-            $processArguments[] = '--max='.$max;
-        }
+        $result = new Binary(file_get_contents($file), $binary->getMimeType(), $binary->getFormat());
 
-        $progressive = \array_key_exists('progressive', $options) ? $options['progressive'] : $this->progressive;
-        if ($progressive) {
-            $processArguments[] = '--all-progressive';
-        } else {
-            $processArguments[] = '--all-normal';
-        }
+        unlink($file);
 
-        $processArguments[] = $input;
-        if ($binary instanceof FileBinaryInterface) {
-            copy($binary->getPath(), $input);
-        } else {
-            file_put_contents($input, $binary->getContent());
+        return $result;
+    }
+
+    /**
+     * @param array $options
+     *
+     * @return array
+     */
+    private function getProcessArguments(array $options = []): array
+    {
+        $arguments = [$this->executablePath];
+
+        if ($options['strip_all'] ?? $this->strip) {
+            $arguments[] = '--strip-all';
         }
 
-        $proc = new Process($processArguments);
-        $proc->run();
+        if (isset($options['max'])) {
+            @trigger_error(sprintf('The "max" option was deprecated in 2.2 and will be removed in 3.0. '.
+                'Instead, use the "quality" option.'), E_USER_DEPRECATED);
 
-        if (false !== mb_strpos($proc->getOutput(), 'ERROR') || 0 !== $proc->getExitCode()) {
-            unlink($input);
-            throw new ProcessFailedException($proc);
+            if (isset($options['quality'])) {
+                throw new InvalidOptionException('the "max" and "quality" options cannot both be set', $options);
+            }
+
+            $options['quality'] = $options['max'];
         }
 
-        $result = new Binary(file_get_contents($input), $binary->getMimeType(), $binary->getFormat());
+        if ($quality = $options['quality'] ?? $this->quality) {
+            if (!in_array($options['quality'], range(0, 100))) {
+                throw new InvalidOptionException('the "quality" option must be an int between 0 and 100', $options);
```

Comment:
would also be nicer if this was discovered in the DI extension instead of at runtime. if this is too generic in the configuration, we could maybe have the processors provide a static validation method that the DI class can call.

Reply:
interesting thought, I'll give it a try

Reply:
as this is not new checks but only refactorings, i think it makes sense to do a separate pull request. is this refactoring ready to be merged in your opinion?